torchao/quantization/quant_api.py

```diff
@@ -199,3 +201,382 @@ def replace_conv2d_1x1(conv):
+        percdamp,
+        groupsize,
+        calibration_tasks,
+        calibration_limit,
+        calibration_seq_length,
+        pad_calibration_inputs,
+        inner_k_tiles=8,
+        padding_allowed=True,
+        precision=torch.float32,
+    ):
+
+        self.tokenizer = tokenizer
+        self.blocksize = blocksize
+        self.percdamp = percdamp
+        self.groupsize = groupsize
+        self.calibration_tasks = calibration_tasks
+        self.calibration_limit = calibration_limit
+        self.calibration_seq_length = calibration_seq_length
+        self.pad_calibration_inputs = pad_calibration_inputs
+        self.inner_k_tiles = inner_k_tiles
+        self.padding_allowed = padding_allowed
+        self.precision = precision
+        self.dyn_quant_func = lambda x: per_token_dynamic_quant(x)
+        n_bit = 4
+        self.get_qparams_func = lambda w: get_group_qparams_symmetric(
+            w, n_bit, groupsize, self.precision
+        )
+        quant_min = -(2 ** (n_bit - 1))
+        quant_max = 2 ** (n_bit - 1) - 1
+        self.quantize_func = lambda w, qparams: torch.ops.quantized_decomposed.quantize_per_channel_group(
+            w, qparams[0], qparams[1], quant_min, quant_max, torch.int8, groupsize
+        )
+        self.dequantize_func = lambda q, qparams: torch.ops.quantized_decomposed.dequantize_per_channel_group(
+            q,
+            qparams[0],
+            qparams[1],
+            quant_min,
+            quant_max,
+            torch.int8,
+            groupsize,
+            self.precision,
+        )
+        self.combine_qparams_list_func = lambda qparams_list: [
+            torch.cat(x, dim=1) for x in zip(*qparams_list)
+        ]
+        # skip unless padding_allowed=True or its correctly sized
+        self.skip_layer_func = lambda linear_weight: not (
+            _check_linear_int4_k(linear_weight.shape[-1], groupsize, inner_k_tiles)
+            or padding_allowed
+        )
+
+        # we need to do the padding here, both for q and the qparams if necessary
+        def make_names_and_values_dict_func(q, qparams):
+            k = q.shape[1]
+            new_k = _calc_padded_size_linear_int4(k, groupsize, inner_k_tiles)
+            # how much we need to pad the weight
+            delta_k = new_k - q.shape[1]
+            final_q = F.pad(q, pad=(0, delta_k))
+            scales_and_zeros = pack_scales_and_zeros(*qparams, precision=self.precision)
+            # how many new groups we need for padded weight
+            delta_groups = new_k // groupsize - scales_and_zeros.shape[0]
+            # TODO: split scales and zero_points
+            final_s_and_z = F.pad(
+                scales_and_zeros, pad=(0, 0, 0, 0, 0, delta_groups), value=1
+            )
+            return {"weight": final_q, "scales_and_zeros": final_s_and_z}
+
+        self.make_names_and_values_dict_func = make_names_and_values_dict_func
+        super().__init__()
+
+    def _convert_for_runtime(self, model):
```

Comment:
Isn't this essentially the meat of it all? Replace all linear layers with this special linear layer. Why use classes at all?

Reply:
this is the meat for replacement, and there is another part for computing the quantized weights in the GPTQ file. the reason that we use classes is to unify the API for all different techniques